File: JavaScript/7js 1/countDevelopers.js
/*
|--------------------------------------------------------------------------
| Coding Meetup #1 - Higher-Order Functions Series - Count the number of JavaScript developers coming from Europe
|--------------------------------------------------------------------------
|
*/

function countDevelopers(list) {
   return list.filter(i => i.continent == 'Europe' && i.language == 'JavaScript').length
}


function countDevelopers(list) {
   return list.reduce((count, dev) =>
      (dev.continent === 'Europe' && dev.language === 'JavaScript') ?
      count + 1 : count, 0);
}


const countDevelopers = list => (
   list.filter(({
      continent,
      language
   }) => continent === 'Europe' && language === 'JavaScript').length
)


function countDevelopers(list) {
   let count = 0;
   for (let i = 0; i < list.length; i++) {
      if (list[i].continent === "Europe" && list[i].language === "JavaScript") {
         count++;
      }
   }
   return count;
}


let list1 = [{
      firstName: 'Noah',
      lastName: 'M.',
      country: 'Switzerland',
      continent: 'Europe',
      age: 19,
      language: 'JavaScript'
   },
   {
      firstName: 'Maia',
      lastName: 'S.',
      country: 'Tahiti',
      continent: 'Oceania',
      age: 28,
      language: 'JavaScript'
   },
   {
      firstName: 'Shufen',
      lastName: 'L.',
      country: 'Taiwan',
      continent: 'Asia',
      age: 35,
      language: 'HTML'
   },
   {
      firstName: 'Sumayah',
      lastName: 'M.',
      country: 'Tajikistan',
      continent: 'Asia',
      age: 30,
      language: 'CSS'
   }
];

console.log(countDevelopers(list1), 1);

File: Python/7py 1/count_developers.py
"""
|--------------------------------------------------------------------------
| Coding Meetup #1 - Higher-Order Functions Series - Count the number of JavaScript developers coming from Europe
|--------------------------------------------------------------------------
|
"""

def count_developers(lst):
   return len(list(filter(lambda i: i['continent'] == 'Europe' and i['language'] == 'JavaScript', lst)))


def count_developers(lst):
   return sum(x["language"] == "JavaScript" and x["continent"] == "Europe" for x in lst)


def count_developers(lst):
   return sum(1 for i in lst if i["continent"] == 'Europe' and i["language"] == 'JavaScript')


def count_developers(lst):
   return sum(dev["language"]=="JavaScript" and dev["continent"]=="Europe" for dev in lst)


list1 = [
{ 'firstName': 'Noah', 'lastName': 'M.', 'country': 'Switzerland', 'continent': 'Europe', 'age': 19, 'language': 'JavaScript' },
{ 'firstName': 'Maia', 'lastName': 'S.', 'country': 'Tahiti', 'continent': 'Oceania', 'age': 28, 'language': 'JavaScript' },
{ 'firstName': 'Shufen', 'lastName': 'L.', 'country': 'Taiwan', 'continent': 'Asia', 'age': 35, 'language': 'HTML' },
{ 'firstName': 'Sumayah', 'lastName': 'M.', 'country': 'Tajikistan', 'continent': 'Asia', 'age': 30, 'language': 'CSS' }
]

print(count_developers(list1))